### PR TITLE
Use boost signals to handle context menu in StatisticIcon

### PR DESCRIPTION
**Pull request by adrianbroher**
Instead of using the GG filter event mechanism to handle the context menu of the various information panels, this PR now uses boost signals and callbacks.  This removes the need to find the associated context menu destination via some id lookup and simplifies the context menu callback.

### CODE REVIEW

File: UI/PopulationPanel.cpp [hunk 1]
```diff
@@ -70,7 +70,30 @@ void PopulationPanel::CompleteConstruction() {
...
-        meter_stat.second->InstallEventFilter(shared_from_this());
+        meter_stat.second->RightClickedSignal.connect([this, &meter_stat](const GG::Pt& pt) {
+            MeterType meter_type = meter_stat.first;
+
+            std::string meter_string = boost::lexical_cast<std::string>(meter_type);
+
+            auto popup = GG::Wnd::Create<CUIPopupMenu>(pt.x, pt.y);
+            auto pc = this->GetPopCenter();
+            if (meter_type == METER_POPULATION && pc) {
+                std::string species_name = pc->SpeciesName();
```
Comment (geoffthemedio) [inline UI/PopulationPanel.cpp:81]: Zoom to species doesn't seem to work in my tests. Quickstarted a game, right click the population icon on my planet and get a two-command popup, pick "Pedia Lookup Human" but nothing happens. Other meter pedia lookup commands do work, though.

File: UI/FleetWnd.cpp [hunk 2]
```diff
@@ -978,7 +977,18 @@ namespace {
             auto icon = GG::Wnd::Create<StatisticIcon>(entry.second, 0, 0, false, StatIconSize().x, StatIconSize().y);
             m_stat_icons.push_back({entry.first, icon});
             AttachChild(icon);
-            icon->InstallEventFilter(shared_from_this());
+            icon->RightClickedSignal.connect([entry](const GG::Pt& pt) {
+                std::string meter_string = boost::lexical_cast<std::string>(entry.first);
+
+                auto zoom_article_action = [&meter_string]() { ClientUI::GetClientUI()->ZoomToMeterTypeArticle(meter_string);};
```
Comment (geoffthemedio) [inline UI/FleetWnd.cpp]: This results in double popup menus from both the icon and the containing fleet panel. First the icon Pedia Lookup X command appears, then after that's closed, a list of fleet commands appears. The latter shouldn't appear if the click was on the icon.

File: UI/FleetWnd.cpp [hunk 3]
```diff
@@ -1842,7 +1756,20 @@ void FleetDataPanel::Init() {
             m_stat_icons.push_back({std::get<0>(entry), icon});
             icon->SetBrowseModeTime(tooltip_delay);
             icon->SetBrowseText(UserString(std::get<2>(entry)));
-            icon->InstallEventFilter(shared_from_this());
+            icon->RightClickedSignal.connect([entry](const GG::Pt& pt){
+                MeterType meter_type = std::get<0>(entry);
+
+                std::string meter_string = boost::lexical_cast<std::string>(meter_type);
+
+                auto zoom_article_action = [&meter_string]() { ClientUI::GetClientUI()->ZoomToMeterTypeArticle(meter_string);};
```
Comment (geoffthemedio) [inline UI/FleetWnd.cpp]: same for this... double popups

File: UI/PopulationPanel.cpp (hunk 1 above)
Comment (geoffthemedio) [reply UI/PopulationPanel.cpp:81]: OK, not sure if this issue is in the PopulationPanel or the MultiIconValueIndiactor. It happens when the population panel is NOT expanded, but it works fine when it is expanded.

Comment (adrianbroher) [reply UI/PopulationPanel.cpp:81]: Cannot reproduce.  Seed 0, Race human, quick start game. German localization.

File: UI/FleetWnd.cpp (hunk 2 above)
Comment (adrianbroher) [reply UI/FleetWnd.cpp]: Confirmed.

File: UI/FleetWnd.cpp (hunk 3 above)
Comment (adrianbroher) [reply UI/FleetWnd.cpp]: Confirmed.

Comment (adrianbroher) [reply UI/FleetWnd.cpp]: Fixed via 4574e50ccff2134536e036a3478bec08c7cfd80e

File: UI/FleetWnd.cpp (hunk 2 above)
Comment (adrianbroher) [reply UI/FleetWnd.cpp]: Fixed via 4574e50ccff2134536e036a3478bec08c7cfd80e

File: UI/PopulationPanel.cpp (hunk 1 above)
Comment (geoffthemedio) [reply UI/PopulationPanel.cpp:81]: To illustrate the issue, I'm trying to use this popup command:

FailedLookupPopup

Switching to the de stringtable doesn't change the result.

Comment (adrianbroher) [reply UI/PopulationPanel.cpp:81]: Still not reproducible, still works as expected.

File: UI/PopulationPanel.cpp [hunk 4]
```diff
@@ -70,7 +70,30 @@ void PopulationPanel::CompleteConstruction() {
...
+                if (!species_name.empty()) {
+                    auto zoom_species_action = [&species_name]() { ClientUI::GetClientUI()->ZoomToSpecies(species_name); };
```
Comment (geoffthemedio) [inline UI/PopulationPanel.cpp]: Seemingly the problem is here, when capturing `species_name` by reference, not by value. It's a local, so is probably out of scope and/or reset to an empty string by a later loop iteration by the time the lambda gets executed.

Removing the `&` from the capture fixes it for me.

Comment (geoffthemedio) [reply UI/PopulationPanel.cpp]: should probably do all the captures in these loops by value...?

Comment (adrianbroher) [reply UI/PopulationPanel.cpp]: Updated PR according to suggestion.

File: UI/PopulationPanel.cpp [hunk 5]
```diff
@@ -70,7 +70,30 @@ void PopulationPanel::CompleteConstruction() {
...
+                std::string species_name = pc->SpeciesName();
+                if (!species_name.empty()) {
+                    auto zoom_species_action = [species_name]() { ClientUI::GetClientUI()->ZoomToSpecies(species_name); };
+                    std::string species_label = boost::io::str(FlexibleFormat(UserString("ENC_LOOKUP")) %
+                                                                              UserString(species_name));
+                    popup->AddMenuItem(GG::MenuItem(species_label, false, false, zoom_species_action));
+                }
+            }
+
+            auto pedia_meter_type_action = [&meter_string]() { ClientUI::GetClientUI()->ZoomToMeterTypeArticle(meter_string); };
```
Comment (geoffthemedio) [inline UI/PopulationPanel.cpp]: still by reference here intentionally?

Comment (adrianbroher) [reply UI/PopulationPanel.cpp]: No, that was an oversight.  Fixed via 17c6807c8d6a026690322d74e173afd3a2a359a3

**Comment by geoffthemedio** [comment]
I think this is causing another issue: drag-dropping onto statistic icons on fleet panels now doesn't work. It's a bit hard to tell because the drag-over highlighting is a bit quirky in master... In master, as far as I can tell, the first statistic icon one drags another fleet panel onto will cause the whole drag-target fleet panel to highlight, but dragging over another statistic icon during the same drag will not. Regardless, dropping seems to still work regardless of whether it's over a statistic icon or the base fleet panel. But with this pull request applied, I can't drop a fleet onto a statistic icon of the target fleet panel; it doesn't execute the drop and fleet merge, and there is also no highlighting of the drop target when dragging over the first (or any) statistic icons of the target fleet panel.

Comparisons in screenshots...
Fleet Panel Statistic Icon Drag Highlighting

**Comment by adrianbroher** [comment]
Bah, the signal delegation in GG sucks.  Added forwarding of drag related signals to StatisticIcon as stop gap for now.

**Pull request merged**